espnet/nets/pytorch_backend/frontends/beamformer.py

```diff
@@ -0,0 +1,165 @@
+import torch
+from torch_complex import functional as FC
+from torch_complex.tensor import ComplexTensor
+
+
+# Ported from https://github.com/fgnt/nn-gev
+
+
+def get_power_spectral_density_matrix(xs: ComplexTensor, mask: torch.Tensor,
+                                      normalization=True,
+                                      eps: float = 1e-15) -> ComplexTensor:
+    """Return cross-channel power spectral density (PSD) matrix
+
+    Args:
+        xs (ComplexTensor): (..., F, C, T)
+        mask (torch.Tensor): (..., F, C, T)
+        normalization (bool):
+    Returns
+        psd (ComplexTensor): (..., F, C, C)
+
+    """
+    # outer product: (..., C_1, T) x (..., C_2, T) -> (..., T, C, C_2)
+    psd_Y = FC.einsum('...ct,...et->...tce', [xs, xs.conj()])
+
+    # Averaging mask along C: (..., C, T) -> (..., T)
+    mask = mask.mean(dim=-2)
+
+    # Normalized mask along T: (..., T)
+    if normalization:
+        # If assuming the tensor is padded with zero, the summation along
+        # the time axis is same regardless of the padding length.
+        mask = mask / (mask.sum(dim=-1, keepdim=True) + eps)
+
+    # psd: (..., T, C, C)
+    psd = psd_Y * mask[..., None, None]
+    # (..., T, C, C) -> (..., C, C)
+    psd = psd.sum(dim=-3)
+
+    return psd
+
+
+def get_mvdr_vector(psd_s: ComplexTensor,
+                    psd_n: ComplexTensor,
+                    reference_vector: torch.Tensor) -> ComplexTensor:
+    """Return the MVDR(Minimum Variance Distortionless Response) vector:
+
+        h = (Npsd^-1 @ Spsd) / (Tr(Npsd^-1 @ Spsd)) @ u
+
+    Reference:
+        On optimal frequency-domain multichannel linear filtering
+        for noise reduction; M. Souden et al., 2010;
+        https://ieeexplore.ieee.org/document/5089420
+
+    Args:
+        psd_s (ComplexTensor): (..., F, C, C)
+        psd_n (ComplexTensor): (..., F, C, C)
+        reference_vector (torch.Tensor): (..., C)
+    Returns:
+        beamform_vector (ComplexTensor)r: (..., F, C)
+    """
+    # numerator: (..., C_1, C_2) x (..., C_2, C_3) -> (..., C_1, C_3)
```

Comment:
I have noticed inverse in pytorch becomes an issue if the magnitude of the entries in the input matrix are very large (may not be really singular) but it throws an error that it is singular. Noticed when using with REVERB data. How about a fix like this:
 
try:                                                                           
        psd_n_i = psd_n.inverse()                                                  
    except:                                                                        
        try:                                                                       
            reg_coeff_tensor = ComplexTensor(torch.rand_like(psd_n.real),          
                                             torch.rand_like(psd_n.real))*1e-2  
            psd_n = psd_n/10e+4                                                    
            psd_s = psd_s/10e+4                                                    
            psd_n += reg_coeff_tensor                                              
            psd_n_i = psd_n.inverse()                                              
        except:                                                                    
            reg_coeff_tensor = ComplexTensor(torch.rand_like(psd_n.real),          
                                             torch.rand_like(psd_n.real))*1e-1  
            psd_n = psd_n/10e+10                                                   
            psd_s = psd_s/10e+10                                                   
            psd_n += reg_coeff_tensor                                              
            psd_n_i = psd_n.inverse()

numerator = FC.einsum('...ec,...cd->...ed', [psd_n_i, psd_s])

Reply:
Can you show me the minimal reproducing script for real matrix inverse? Perhaps, my complex tensor has a problem.If not so, we should launch a issue in pytorch.

Reply:
I am not able to find a simple example for real, but look at the simple 2*2 example with your complex tensor that I encountered while running with REVERB data. 
![Screen Shot 2019-04-12 at 12 50 34 PM](https://user-images.githubusercontent.com/26885951/56053258-a8466380-5d21-11e9-9332-1d5e09d177c0.png)

It gives an error in this line: o_real = (a.real + torch.matmul(torch.matmul(a.imag, a.real.inverse()), a.imag)).inverse(), where you have used the real inverse of pytorch, so I guess that has the same issue too.

Reply:
I was able to take the inverse of the same matrix by just dividing by a constant.
![Screen Shot 2019-04-12 at 12 56 22 PM](https://user-images.githubusercontent.com/26885951/56053576-7e417100-5d22-11e9-9697-158b088c34bb.png)

that is why in my fix I divide both psd_s and psd_n by a constant if the original inverse fails, as you take inverse of one and multiply both, I guess it won't make any difference in the final solution
reg_coeff_tensor = ComplexTensor(torch.rand_like(psd_n.real),
torch.rand_like(psd_n.real))*1e-2
psd_n = psd_n/10e+4
psd_s = psd_s/10e+4
psd_n += reg_coeff_tensor
I also add a small constant in case if the matrix is really singular. I have two except blocks as if division by one constant fails, I try a a constant with higher magnitude.

Reply:
Thanks. In your example, real singular matrix seems to happen. So this is a problem of ComlexTensor. Let me think about this meanwhile.

https://github.com/kamo-naoyuki/pytorch_complex/blob/2673fdc6bae3e74be2828283e48dd5f6b39d0c5a/torch_complex/tensor.py#L378-L379

```
tensor([[[2.6844e+08, 2.6844e+08],
         [2.6844e+08, 2.6844e+08]]])
```

Reply:
I updated pytorch_complex to try inverse() with several `t` parameters.
(pytorch_complex doesn't invoke the inverse of blas API directly, but contaminates the torch.Tensor.invserse() to describe the inverse() of complex matirx referring https://www.math.uwaterloo.ca/~hwolkowi/matrixcookbook.pdf  )

The condition number of this matrix looks too large, so it is hard to get the inverse() accurately essentially. 

```python
import numpy
import torch
from torch_complex.tensor import ComplexTensor

real = torch.tensor([[1.43380048e+15, 1.11050124e+15],
                     [1.11050124e+15, 1.11960107e+15]])
imag = torch.tensor([[8.758782e-04, 6.099766e+14],
                     [-6.099766e+14, 9.679516e-04]])
a = ComplexTensor(real, imag)
print(a.inverse() @ a)

an = a.numpy()
print(numpy.linalg.inv(an) @ an)
print('cond =', numpy.linalg.cond(an))
```

```
ComplexTensor(
Real:
tensor([[2.0000, 0.2500],
        [0.0000, 1.2500]])
Imag:
tensor([[ 2.2500,  2.2500],
        [-3.2500, -2.4375]])
)
[[ 0.        +1.j -0.91687727+0.j]
 [ 1.3755432 +0.j  2.        -1.j]]
cond = 116339144.0
```

Perhaps we may want to skip the uptating in training in this case, but I think we can ignore it because this is very rare event. 
I'm not particular about numerical calculation so much,  so what do you think?

Reply:
Yes maybe for now at least we should include code to skip updating. I think we should not ignore it as it not that rare. I encounter it quite often while running with REVERB data and also in the correlation matrix inverse part of WPE. It will be nice if we can find a permanent fix to it.

Reply:
WPE and REVERB is not within the range of this PR, so I won't include for now. It requires several changing in the main codes. 

Maybe such unstable matrix often happens in early training steps and that is not dominant as a whole, so , at least, the training is possible in the current code.

Reply:
OK. We can just make it pending for this PR.

Reply:
@sas91 Maybe I don't have enough time  for the duration after this PR. If you could provide your base script of REVERB as another PR, it's great for me.
Thanks for your help for this.